compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -743,6 +753,117 @@ fn find_opaque_ty_constraints_for_tait(tcx: TyCtxt<'_>, def_id: LocalDefId) -> T
+fn may_define_opaque_type<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    opaque_def_id: LocalDefId,
+) -> bool {
+    if tcx.is_descendant_of(opaque_def_id.to_def_id(), def_id.to_def_id()) {
+        // If the opaque type is defined in the body of a function, that function
+        // may constrain the opaque type since it can't mention it in bounds.
+        return true;
+    }
+
+    if tcx.is_closure(def_id.to_def_id()) {
+        return may_define_opaque_type(tcx, tcx.local_parent(def_id), opaque_def_id);
+    }
+
+    let param_env = tcx.param_env(def_id);
+
+    trace!(parent = ?tcx.parent(opaque_def_id.to_def_id()));
+    fn has_tait<'tcx>(
+        val: impl TypeVisitable<TyCtxt<'tcx>>,
+        opaque_def_id: LocalDefId,
+        tcx: TyCtxt<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+    ) -> bool {
+        struct Visitor<'tcx> {
+            opaque_def_id: DefId,
+            tcx: TyCtxt<'tcx>,
+            seen: FxHashSet<Ty<'tcx>>,
+            param_env: ty::ParamEnv<'tcx>,
+        }
+        impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for Visitor<'tcx> {
+            type BreakTy = ();
+
+            #[instrument(skip(self), level = "trace", ret)]
+            fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
+                // Erase all lifetimes, they can't affect anything, but recursion
+                // may cause late bound regions to differ, because the type visitor
+                // can't erase them in `visit_binder`.
+                let t = t.fold_with(&mut ty::fold::BottomUpFolder {
+                    tcx: self.tcx,
+                    ct_op: |c| c,
+                    ty_op: |t| t,
+                    lt_op: |_| self.tcx.lifetimes.re_erased,
+                });
+                if !self.seen.insert(t) {
+                    return ControlFlow::Continue(());
+                }
+                match t.kind() {
+                    ty::Alias(ty::Opaque, alias) => {
+                        if alias.def_id == self.opaque_def_id {
+                            return ControlFlow::Break(());
+                        }
+                        for (pred, _span) in self
+                            .tcx
+                            .bound_explicit_item_bounds(alias.def_id)
+                            .subst_iter_copied(self.tcx, alias.substs)
+                        {
+                            pred.visit_with(self)?;
+                        }
+                    }
+                    ty::Alias(ty::Projection, _) => {
+                        if let Ok(proj) = self.tcx.try_normalize_erasing_regions(self.param_env, t)
+                        {
+                            proj.visit_with(self)?;
+                        }
+                    }
+                    // Types that have opaque type fields must get walked manually, they
+                    // would not be seen by the type visitor otherwise.
+                    ty::Adt(adt_def, substs) => {
```

Comment:
so the type can be contained in an `Adt`? please add that to the PR description.

also, this can be `adt_def.all_fields` instead of the variants -> fields nested loop